File: iOS_SDK/OneSignal/OneSignalTrackIAP.m
/**
 * Modified MIT License
 *
 * Copyright 2015 OneSignal
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * 1. The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * 2. All copies of substantial portions of the Software may only be used in connection
 * with services provided by OneSignal.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */

#import <Foundation/Foundation.h>
#import <objc/runtime.h>
#import "OneSignalTrackIAP.h"
#import "OneSignal.h"

@implementation OneSignalTrackIAP

// NSClassFromString and performSelector are used so OneSignal does not depend on StoreKit to link the app.

static Class skPaymentQueue;
NSMutableDictionary* skusToTrack;

+ (BOOL)canTrack {
    skPaymentQueue = NSClassFromString(@"SKPaymentQueue");
    return (skPaymentQueue != nil && [skPaymentQueue performSelector:@selector(canMakePayments)]);
}

- (id)init {
    self = [super init];
    
    if (self)
        [[skPaymentQueue performSelector:@selector(defaultQueue)] performSelector:@selector(addTransactionObserver:) withObject:self];
    
    return self;
}

- (void)paymentQueue:(id)queue updatedTransactions:(NSArray*)transactions {
    skusToTrack = [NSMutableDictionary new];
    id skPayment;
    
    for (id transaction in transactions) {
        NSInteger state = [transaction performSelector:@selector(transactionState)];
        switch (state) {
            case 1: // SKPaymentTransactionStatePurchased
                skPayment = [transaction performSelector:@selector(payment)];
                NSString* sku = [skPayment performSelector:@selector(productIdentifier)];
                NSInteger quantity = [skPayment performSelector:@selector(quantity)];
                
                if (skusToTrack[sku])
                    [skusToTrack[sku] setObject:[NSNumber numberWithInt:[skusToTrack[sku][@"count"] intValue] + quantity] forKey:@"count"];
                else
                    skusToTrack[sku] = [NSMutableDictionary dictionaryWithObjectsAndKeys:[NSNumber numberWithInteger:quantity], @"count", nil];
                break;
        }
    }
    
    if (skusToTrack.count > 0)
        [self getProductInfo:[skusToTrack allKeys]];
}


- (void)getProductInfo:(NSArray*)productIdentifiers {
    Class SKProductsRequestClass = NSClassFromString(@"SKProductsRequest");
    id productsRequest = [[SKProductsRequestClass alloc]
                            performSelector:@selector(initWithProductIdentifiers:) withObject:[NSSet setWithArray:productIdentifiers]];
    [productsRequest setDelegate:self];
    [productsRequest performSelector:NSSelectorFromString(@"start")];
}

- (void)productsRequest:(id)request didReceiveResponse:(id)response {
    NSMutableArray* arrayOfPruchases = [NSMutableArray new];
    
    for(id skProduct in [response performSelector:@selector(products)]) {
        NSString* productSku = [skProduct performSelector:@selector(productIdentifier)];
        NSMutableDictionary* purchase = skusToTrack[productSku];
        if (purchase) { // In rare cases this can be nil when there wasn't a connection to Apple when opening the app but there was when buying an IAP item.
            purchase[@"sku"] = productSku;
            purchase[@"amount"] = [skProduct performSelector:@selector(price)];
            purchase[@"iso"] = [[skProduct performSelector:@selector(priceLocale)] objectForKey:NSLocaleCurrencyCode];
            if ([purchase[@"count"] intValue] == 1)
                [purchase removeObjectForKey:@"count"];
            [arrayOfPruchases addObject:purchase];
        }
    }
    
    if ([arrayOfPruchases count] > 0)
        [[OneSignal defaultClient] performSelector:@selector(sendPurchases:) withObject:arrayOfPruchases];
}


@end

File: OneSignalSwift/OneSignalSwift/OneSignalTrackIAP.swift
//
//  OneSignalTrackIAP.swift
//  OneSignalSwift
//
//  Created by Joseph Kalash on 6/21/16.
//  Copyright © 2016 OneSignal. All rights reserved.
//

import Foundation

class OneSignalTrackIAP : NSObject {
    
    static var skPaymentQueue : AnyClass?
    var skusToTrack : NSMutableDictionary?
    
    class func canTrack () -> Bool {
        
        skPaymentQueue = NSClassFromString("SKPaymentQueue")
        if let nsobject = OneSignalTrackIAP.skPaymentQueue as? NSObjectProtocol {
            if nsobject.performSelector(NSSelectorFromString("canMakePayments")).takeUnretainedValue() as? Bool == true { return true }
        }
        
        return false
    }
    
    
    override init() {
        
        super.init()
        
        let defaultQueue = (OneSignalTrackIAP.skPaymentQueue as? NSObjectProtocol)?.performSelector(NSSelectorFromString("defaultQueue"))
        (defaultQueue as? NSObjectProtocol)?.performSelector(NSSelectorFromString("addTransactionObserver:"), withObject: self)
    }
    
    func paymentQueue(queue: AnyClass, updatedTransactions transactions : NSArray) {
        skusToTrack = NSMutableDictionary()
        var skPayment : AnyObject?
        
        for transaction in transactions {
            if let state = transaction.valueForKey("transactionState") as? NSInteger where state == 1 {
                skPayment = transaction.valueForKey("payment")
                let sku = skPayment?.performSelector(NSSelectorFromString("productIdentifier")).takeUnretainedValue() as? NSString
                var qty : Int32 = 0
                if let q = skPayment?.valueForKey("quantity") as? NSInteger { qty = Int32(q) }
                if sku != nil {
                    if let sku_dict = skusToTrack![sku!] as? NSMutableDictionary {
                        var oldCount : Int32 = 0
                        if let old = skusToTrack![sku!]?["count"] as? NSNumber { oldCount = old.intValue }
                        sku_dict.setObject(NSNumber(int: (qty + oldCount)), forKey: "count")
                    }
                    else { skusToTrack![sku!] = NSMutableDictionary(object: NSNumber(int: qty), forKey: "count") }
                }
            }
        }
        
        if skusToTrack!.count > 0 { getProductInfo(skusToTrack!.allKeys)}
    }
    
    func getProductInfo(productIdentifiers: NSArray) {
        
        let SKProductsRequestClass : AnyClass? = NSClassFromString("SKProductsRequest")
        let productsRequest : AnyObject? = SKProductsRequestClass?.alloc()
        (productsRequest as? NSObjectProtocol)?.performSelector(NSSelectorFromString("initWithProductIdentifiers:"), withObject: NSSet(array: productIdentifiers as [AnyObject]))
        productsRequest?.setValue(self, forKey: "delegate")
        (productsRequest as? NSObjectProtocol)?.performSelector(NSSelectorFromString("start"))
        
    }
    
    
    func productsRequest(request: NSObject, didReceiveResponse response: NSObject) {
        let arrayOfPurchases = NSMutableArray()
        
        if let products = response.valueForKey("products") as? NSArray {
            for product in products {
                if let sku = product.valueForKey("productIdentifier") as? NSString, purchase = skusToTrack![sku] as? NSMutableDictionary {
                    purchase["sku"] = sku
                    purchase["amount"] = product.valueForKey("price") as? NSDecimalNumber
                    purchase["iso"] = (product.valueForKey("priceLocale") as? NSLocale)?.objectForKey(NSLocaleCurrencyCode)
                    if let count = purchase["count"] as? NSNumber where count.intValue == 1 {
                        purchase.removeObjectForKey("count")
                    }
                    arrayOfPurchases.addObject(purchase)
                }
            }
            
            if arrayOfPurchases.count > 0 {
                OneSignal.defaultClient.performSelector(#selector(OneSignal.sendPurchases(_:)), withObject: arrayOfPurchases)
            }
        }
    }
    
    
}
